### PR TITLE
4926314: Optimize Reader.read(CharBuffer)

### PR DESCRIPTION
**Pull request by marschall**
Implement three optimiztations for Reader.read(CharBuffer)

* Add a code path for heap buffers in Reader#read to use the backing array instead of allocating a new one.
* Change the code path for direct buffers in Reader#read to limit the intermediate allocation to `TRANSFER_BUFFER_SIZE`.
* Implement `InputStreamReader#read(CharBuffer)` and delegate to `StreamDecoder`.
* Implement `StreamDecoder#read(CharBuffer)` and avoid buffer allocation.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-4926314](https://bugs.openjdk.java.net/browse/JDK-4926314): Optimize Reader.read(CharBuffer)

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to c4c859e075f0b9bb21077b872e7719453df97e94
 * [Brian Burkhalter](https://openjdk.java.net/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/1915/head:pull/1915` \
`$ git checkout pull/1915`

Update a local copy of the PR: \
`$ git checkout pull/1915` \
`$ git pull https://git.openjdk.java.net/jdk pull/1915/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1915`

View PR using the GUI difftool: \
`$ git pr show -t 1915`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/1915.diff">https://git.openjdk.java.net/jdk/pull/1915.diff</a>

</details>

### CODE REVIEW

**Comment by marschall** [comment]
A couple of implementation notes:

Reader#read(CharBuffer)

on-heap case

I introduced a dedicated path for the on-heap case and directly read into the backing array. This completely avoids the intermediate allocation and copy. Compared to the initial proposal the buffer position is updated. This has to be done because we bypass the buffer and directly read into the array. This also handles the case that #read returns -1.

I am using a c-style array declaration because the rest of the class uses it.

off-heap case

I limit the intermadiate allocation to TRANSFER_BUFFER_SIZE. In order to reduce the total intermediate allocation we now call #read multiple times until the buffer is full or EOF is reached. This changes the behavior slightly as now possibly more data is read. However this should contribute to reduce the overall intermediate allocations.

A lock is acquired to keep the the read atomic. This is consistent with #skip which also holds a lock over multiple #read calls. This is inconsistent with #transferTo which does not acquire a lock over multiple #read calls. 

The implementation took inspiration from java.io.InputStream#readNBytes(int).

InputStreamReader#read(CharBuffer)

Since StreamDecoder is a Reader as well we can simply delegate.

StreamDecoder#read(CharBuffer)

Interestingly this was not implemented even though StreamDecoder internally works on NIO buffers.

on-heap case

We see a performance improvement and the elimination of all intermediate allocation.

StreamDecoder#read(char[], int, int) and InputStreamReader#read(char[], int, int) may get a small improvement as we no longer #slice the buffer.

off-heap case

We see the elimination of all intermediate allocation but a performance penalty because we hit the slow path in #decodeLoop. There is a trade-off here, we could improve the performance to previous levels by introducing intermediate allocation again. I don't know how much the users of off-heap buffers care about introducing intermediate allocation vs maximum throughput.

I was struggling to come up with microbenchmarks because the built in InputStream and Reader implementations are finite but JMH calls the benchmark methods repeatably. I ended up implementing custom infinite InputStream and Reader implementations. The code can be found there:

https://github.com/marschall/reader-benchmarks/tree/master/src/main/java/com/github/marschall/readerbenchmarks

An Excel with charts can be found here:

https://github.com/marschall/reader-benchmarks/raw/master/src/main/resources/4926314.xlsx

I looked for java.io.Reader#read(CharBuffer) users in the JDK and only found java.util.Scanner#readInput(). I wrote a microbenchmark for this but only found minuscule improvements due to regex dominating the runtime.

There seem to be no direct Reader tests in the tier1 suite, the initial code was wrong because I forgot to update the buffer code position but I only found out because some Scanner tests failed.

**Comment by bplb** [comment]
I changed the JBS issue summary to match the title of this PR so that integration blocker is removed.

How does the performance of `InputStreamReader.read(CharBuffer)` compare for the case where only the change to `Reader` is made versus if all your proposed changes are applied?

Note that the more recent copyright year is now 2021.

File: src/java.base/share/classes/java/io/Reader.java
Comment (bokken) [inline src/java.base/share/classes/java/io/Reader.java:198]: Would there be value in making this a (lazily created) member variable? That would allow a single instance to be reused. It seems likely that, if one call is made with a direct CharBuffer, subsequent calls will also be made with direct instances (probably same instance?).

Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java:198]: I am not sure. It would help to reduce the allocation rate when reading a large amount of data into a small direct CharBuffer. I don't know how common that is. We would introduce an instance variable for one path in one method.

File: src/java.base/share/classes/java/io/Reader.java
Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java:207]: You're correct. I need to expand the unit tests.

Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java:206]: Fixed

File: src/java.base/share/classes/java/io/Reader.java
Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java:197]: That would be possible. It would help in cases where a large Reader is read into one or several relatively small off-heap CharBuffers, requiring multiple #read calls. This can only be done when the caller is able to work with only a partial input. I don't know how common this case is.

We could re-purpose #skipBuffer, it has the same maximum size (8192) but determined by a different constant (#maxSkipBufferSize instead of #TRANSFER_BUFFER_SIZE). That would likely require it to be renamed and maybe we should even remove #maxSkipBufferSize. We could also do the reallocation and growing similar as is currently done in #skip.

File: src/java.base/share/classes/java/io/Reader.java
Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java]: Done

File: src/java.base/share/classes/java/io/Reader.java
Comment (bplb) [reply src/java.base/share/classes/java/io/Reader.java:200]: Perhaps a static final `WORK_BUFFER_SIZE` could be added with value 8192 and `maxSkipBufferSize` and `TRANSFER_BUFFER_SIZE` replaced with that? Then `skipBuffer` could be renamed to `workBuffer` and used in both `read(CharBuffer)` and `skip(long)`. That shouldn't be a problem as both uses are in synchronized blocks. Also I suggest putting the declaration of `workBuffer` just below that of `lock` instead of lower down the file where `skipBuffer` is.

Lastly you mentioned C-style array declarations like `char buf[]`. As there are only four of these in the file it might be good to just go ahead and change them, I don't think that adds much noise or risk.

Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java]: Done. I left #transferTo(Writer) untouched for now. Firstly it is not already behind a synchronized. Secondly it writes so there is no need for repeated calls.

**Comment by bplb** [comment]
I think the implementation changes here look good. I don't know however whether there is enough coverage in the tests. These should verify that the `Reader`, `CharArrayReader`, and `InputStreamReader` implementations of `read(CharBuffer)` are accurate. If there is already sufficient coverage in the tests in `test/jdk/java/io` then that is good enough and nothing need be added.

File: test/jdk/java/io/InputStreamReader/ReadCharBuffer.java
Comment (bplb) [inline test/jdk/java/io/InputStreamReader/ReadCharBuffer.java:34]: It's generally better not to use a wild card.

File: test/jdk/java/io/InputStreamReader/ReadCharBuffer.java
Comment (bplb) [inline test/jdk/java/io/InputStreamReader/ReadCharBuffer.java:73]: I think `buffer.rewind()` would be more in keeping with the specification verbiage although there will be no practical effect here. Same comment applies below and in the other test `ReadCharBuffer`.

File: src/java.base/share/classes/java/io/Reader.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/Reader.java:221]: The code for case that the char buffer has a backing array looks okay but I'm not sure about the direct buffer/other cases. One concern is that this is a read method, not a transferXXX method so we shouldn't be calling the underlying read several times. You'll see what I mean if you consider the scenario where you read < rem, then read again and the second read blocks or throws. I'l also concerned about "workBuffer" adding more per-stream footprint for cases where skip or read(CB) is used. Objects such as InputStreamReader are already a problem due to the underlying stream decoder.

Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java:221]: Right. So you propose to revert the off-heap path to the current master? That would be fine with me. The original bug and my motivation was only about the backing array case, the rest crept in. That would certainly keep the risk and impact lower.

Comment (bplb) [reply src/java.base/share/classes/java/io/Reader.java:221]: I think that's what @AlanBateman intended. The `skip()` changes would revert also but the C-style array changes can stay. Thanks.

Comment (AlanBateman) [reply src/java.base/share/classes/java/io/Reader.java:221]: Yes, let's keep bring it back to just eliminating the intermediate array when the buffer has a backing array. The other case that been examined separated if needed but we can't use the approach proposed in the current PR because it changes the semantics of read when there is a short-read followed by a blocking or exception throwing read.

File: test/jdk/java/io/InputStreamReader/ReadCharBuffer.java
Comment (marschall) [reply test/jdk/java/io/InputStreamReader/ReadCharBuffer.java]: Done

File: test/jdk/java/io/InputStreamReader/ReadCharBuffer.java
Comment (marschall) [reply test/jdk/java/io/InputStreamReader/ReadCharBuffer.java:77]: `buffer.rewind()` would not work in this specific case as it does not reset the limit. I want to assert the entire buffers content to make sure the method didn't accidentally write where it shouldn't have. Therefore limit needs to be set to capacity.

File: src/java.base/share/classes/java/io/Reader.java
Comment (marschall) [reply src/java.base/share/classes/java/io/Reader.java]: Done

**Comment by marschall** [comment]
> I think the implementation changes here look good. I don't know however whether there is enough coverage in the tests. These should verify that the `Reader`, `CharArrayReader`, and `InputStreamReader` implementations of `read(CharBuffer)` are accurate. If there is already sufficient coverage in the tests in `test/jdk/java/io` then that is good enough and nothing need be added.

`CharArrayReader` was lacking a test. I added a test which found a bug and fixed the bug. The PR also contains new tests for `Reader` and `InputStreamReader`. They cover on-heap and off-heap cases.

Is there a way to get test coverage with JTReg tests? I only found [1] which seems out of date and points to an Oracle internal wiki.

 [1] https://wiki.openjdk.java.net/display/CodeTools/JCov+FAQ#JCovFAQ-HowdoIuseJCovwithjtreg?

File: src/java.base/share/classes/java/io/Reader.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/Reader.java:205]: Thanks for bringing this back to just the heap buffer case. This part looks good now.

Comment (bplb) [reply src/java.base/share/classes/java/io/Reader.java:205]: Agreed.

**Comment by marschall** [comment]
How do we proceed here? Are there additional changes that you would like me to perform or undo?

File: test/jdk/java/io/CharArrayReader/ReadCharBuffer.java
Comment (bplb) [inline test/jdk/java/io/CharArrayReader/ReadCharBuffer.java:50]: "as they may"

File: test/jdk/java/io/InputStreamReader/ReadCharBuffer.java
Comment (bplb) [inline test/jdk/java/io/InputStreamReader/ReadCharBuffer.java:52]: "as they may"

File: test/jdk/java/io/Reader/ReadCharBuffer.java
Comment (bplb) [inline test/jdk/java/io/Reader/ReadCharBuffer.java:51]: "as they may"

File: test/jdk/java/io/Reader/ReadCharBuffer.java
Comment (bplb) [inline test/jdk/java/io/Reader/ReadCharBuffer.java:83]: Lines 80 + 83 might be easier to read if replaced with
```
            int position = 8;
            limit = position + 8192 + 1;
            buffer.limit(limit);
            buffer.position(position);
            assertEquals(reader.read(buffer), limit - position);
```

**Review by AlanBateman — APPROVED** [review]
Thanks for rolling back to the direct buffer case, the updated implementation changes look okay.

**Review by bplb — APPROVED** [review]
Approved modulo my "as they may" remarks on three comment lines in the tests.

**Comment by bplb** [comment]
This PR is on hold pending the author's issuing the `/integrate` command.

File: test/jdk/java/io/CharArrayReader/ReadCharBuffer.java
Comment (marschall) [reply test/jdk/java/io/CharArrayReader/ReadCharBuffer.java]: Done, and the other two as well.

**Review by bplb — APPROVED** [review]

**Pull request closed without merging**
